### ISSUE TITLE
Consumers getting stuck with "Reached max fetcher queue size (10), sleeping 1s" warning

### ISSUE THREAD

**Issue by Azdaroth** [zendesk/ruby-kafka#787]
* Version of Ruby: 2.6
* Version of Kafka: 1.1
* Version of ruby-kafka: 0.7.10

I have a problem for large topics where after a while, consumers (from Karafka gem 1.3.0) get stuck with "Reached max fetcher queue size (10), sleeping 1s" warning. And this is quite random depending on the application (or rather how topics/partitions are assigned to consumers). So some workers don't have any problems at all and some of them are stuck, sometimes all of them are stuck and scaling up/down workers doesn't help. 

What would be the way to unstuck the consumers and why does it happen? The topics are indeed large but after a day or two, the consumers would be able to catch up with the offset lag and get to the beginning of the topic so it's not the case that more messages are published that the consumers would be ever able to consume. Would changing a config for max fetcher queue size help and if yes, what would be the recommended value? And is it actually a feature or a bug that the consumers can get stuck with just a warning and not an exception for example? 

If it somehow helps, the topics are log-compacted, there are 5 topics and each of them has 10 partitions. We tried various numbers of workers between 2 and 30.

**Comment by Azdaroth** [comment]
I tried different values for max fetcher queue size 2 and 100 and seems like it's the same case for all of them.

**Comment by mensfeld** [comment]
So, the "Reached max fetcher queue size" is nothing abnormal if you have a topic with a huge lag.

Also, please define what "for large topics" mean.Huge messages or huge amount of messages.

Keep in mind, that if you subscribe to a topic that has a lot of messages (and new being produced) and you end up consuming slower than you produce, weird things may happen if you hit the retency policy application.

Also: can you provide logs? I mean debug logs (you can set both ruby-kafka and karafka at the same time).

> Would changing a config for max fetcher queue size help and if yes, what would be the recommended value?

Recommended here depends on size of your machine and it's power. It depends.

>  is it actually a feature or a bug that the consumers can get stuck with just a warning and not an exception for example?

We don't know yet whether it is a bug or a side-effect of you doing something wrong ;)

**Comment by Azdaroth** [comment]
@mensfeld Thanks for the response. 

The retention policy is set to millions of years, so that shouldn't cause any issue, despite the topic being huge in terms of number of messages. Seems like one of the consumer group has caught up eventually with all the messages, so maybe "getting stuck" was just temporary. I'll also try to use Sidekiq backend for Karafka as this will surely help consume faster.

If that won't help, I'll provide some logs, don't want to make too much noise here if there are still some options to try.

**Comment by Azdaroth** [comment]
This issue can be closed, I've managed to solve the problem. Apparently, it was fully on the application's side and literally it was about missing indexes in one app that were present in the other one and that was the source of the major difference in behavior between the apps 🤦‍♂ One topic had some heavy queries performed from time to time and this is when the processing was really slow (3-4s per message or even more) + it was causing a massive DB load due to sequential parallel scans which was also causing other topics to be slow. So in the end, consumers were getting kicked out from Kafka all the time (I guess after 30s which was the limit from the config) and that's why they appeared to be stuck.

Thanks for help, and hope it helps someone else when they encounter this kind of problem, I also found that decreasing max fetcher queue size helps a bit with getting unstuck, but certainly, that is not a solution here :P.
